solr/core/src/java/org/apache/solr/core/TransientSolrCoreCacheDefault.java

```diff
@@ -18,103 +18,119 @@
   }
-  // This just moves the 
-  private void doInit() {
-    NodeConfig cfg = coreContainer.getNodeConfig();
+
+  private int getConfiguredCacheMaxSize(CoreContainer container) {
+    int configuredCacheMaxSize = NodeConfig.NodeConfigBuilder.DEFAULT_TRANSIENT_CACHE_SIZE;
+    NodeConfig cfg = container.getNodeConfig();
     if (cfg.getTransientCachePluginInfo() == null) {
-      // Still handle just having transientCacheSize defined in the body of solr.xml not in a transient handler clause.
-      this.cacheSize = cfg.getTransientCacheSize();
+      // Still handle just having transientCacheSize defined in the body of solr.xml
+      // not in a transient handler clause.
+      configuredCacheMaxSize = cfg.getTransientCacheSize();
     } else {
       @SuppressWarnings({"rawtypes"})
       NamedList args = cfg.getTransientCachePluginInfo().initArgs;
       Object obj = args.get("transientCacheSize");
       if (obj != null) {
-        this.cacheSize = (int) obj;
+        configuredCacheMaxSize = (int) obj;
       }
     }
-
-    // it's possible for cache
-    if (cacheSize < 0) { // Trap old flag
-      cacheSize = Integer.MAX_VALUE;
+    if (configuredCacheMaxSize < 0) { // Trap old flag
+      configuredCacheMaxSize = Integer.MAX_VALUE;
     }
-
-    // Now don't allow ridiculous allocations here, if the size is > 1,000, we'll just deal with
-    // adding cores as they're opened. This blows up with the marker value of -1.
-    int actualCacheSize = Math.min(cacheSize, 1000);
-    log.info("Allocating transient cache for {} transient cores", actualCacheSize);
-    transientCores = new LinkedHashMap<>(actualCacheSize, 0.75f, true) {
-      @Override
-      protected boolean removeEldestEntry(Map.Entry<String, SolrCore> eldest) {
-        if (size() > cacheSize) {
-          SolrCore coreToClose = eldest.getValue();
-          if (log.isInfoEnabled()) {
-            log.info("Closing transient core [{}]", coreToClose.getName());
-          }
-          coreContainer.queueCoreToClose(coreToClose);
-          return true;
-        }
-        return false;
-      }
-    };
+    return configuredCacheMaxSize;
   }
 
-  
   @Override
   public Collection<SolrCore> prepareForShutdown() {
-    // Return a copy of the values
-
-    @SuppressWarnings({"unchecked", "rawtypes"})
-    List<SolrCore> ret = new ArrayList(transientCores.values());
-    transientCores.clear();
+    // Return a copy of the values.
+    List<SolrCore> ret = new ArrayList<>(transientCores.asMap().values());
+    transientCores.invalidateAll();
+    transientCores.cleanUp();
     return ret;
   }
 
   @Override
-  public CoreContainer getContainer() { return this.coreContainer; }
+  public CoreContainer getContainer() {
+    return coreContainer;
+  }
 
   @Override
   public SolrCore addCore(String name, SolrCore core) {
-    return transientCores.put(name, core);
+    return transientCores.asMap().put(name, core);
```

Comment:
I'm unclear why we need the intermediate `asMap` call here and in other methods.

Reply:
The Cache interface is more opinionated regarding common expectations of usage patterns, so `cache.put` returns `void` whereas `Map.put` returns the old value. Generally a cache isn't inspected, e.g. there is no iterator of memcached's entries. The `asMap` provides a more advanced escape hatch because an opinionated interface doesn't match developer needs, so we just nudge towards the usual best practices.

Reply:
Thanks Ben for reviewing!
Yes I used asMap() for the following reasons:
- The TransientSolrCoreCache interface requires to return the old SolrCore associated to the name, which is only available with the asMap() view.
- The iterators on other methods are also only available through the asMap() view.
- There is no performance penalty to use the asMap() view. It's just the way the Cache interface is separated between pure Cache needs and Map-like needs.
However there is a slight difference in the way cache stats are recorded. That's why I added a javadoc comment in the transientCore field.